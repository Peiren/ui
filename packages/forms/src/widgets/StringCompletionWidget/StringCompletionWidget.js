import React, { PropTypes } from 'react';
import Typeahead from 'react-talend-components/lib/Typeahead';

function handleSelect(items, onChange) {
	return function onSelect(event, { sectionIndex, itemIndex }) {
		onChange(items[sectionIndex].suggestions[itemIndex].title);
	};
}

function StringCompletionWidget(props) {
	let items = [];
	if (props.options.itemsSrc) {
		items = props.formContext.fetchItems(props.options.itemsSrc);
	}
	return (
		<Typeahead
			items={items}
			value={props.value || ''}
			onChange={change => props.onChange(change.target.value)}
			onSelect={handleSelect(items, props.onChange)}
		/>
	);
}

if (process.env.NODE_ENV !== 'production') {
	StringCompletionWidget.propTypes = {
		options: PropTypes.shape({
			itemsSrc: PropTypes.string,
		}).required,
		value: PropTypes.string,
		onChange: PropTypes.func,
		formContext: PropTypes.shape({
			fetchItems: PropTypes.func.required,
		}),
	};
}

export default StringCompletionWidget;
