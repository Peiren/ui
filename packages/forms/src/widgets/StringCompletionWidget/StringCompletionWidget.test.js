import React from 'react';
import { shallow } from 'enzyme';
import StringCompletionWidget from './StringCompletionWidget';

function fetchItems() {
	return [{
		title: 'A',
		suggestions: [{
			title: 'Apple',
		}, {
			title: 'Apricot',
		}],
	}, {
		title: 'B',
		suggestions: [{
			title: 'Banana',
		}],
	}, {
		title: 'C',
		suggestions: [{
			title: 'Cherry',
		}],
	}];
}

const options = { itemsSrc: '/test.json' };

describe('StringCompletionWidget', () => {
	it('should render', () => {
		const wrapper = shallow(
			<StringCompletionWidget options={options} fetchItems={fetchItems} />
		);
		expect(wrapper.prop('items')).toBeDefined();
		expect(wrapper).toMatchSnapshot();
	});

	it('should render without items', () => {
		const wrapper = shallow(
			<StringCompletionWidget options={{}} fetchItems={() => {}} />
		);
		expect(wrapper.prop('items')).toEqual([]);
		expect(wrapper).toMatchSnapshot();
	});

	it('should call fetchItems', () => {
		const fi = jest.fn();
		shallow(
			<StringCompletionWidget options={options} fetchItems={fi} />
		);
		expect(fi.mock.calls[0][0]).toBeDefined();
		expect(fi.mock.calls.length).toBe(1);
	});
});
