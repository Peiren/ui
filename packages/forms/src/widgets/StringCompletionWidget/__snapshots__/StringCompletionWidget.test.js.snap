exports[`StringCompletionWidget should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Typeahead
    id={42}
    items={
        Array [
            Object {
              "suggestions": Array [
                Object {
                  "title": "Apple",
                },
                Object {
                  "title": "Apricot",
                },
              ],
              "title": "A",
            },
            Object {
              "suggestions": Array [
                Object {
                  "title": "Banana",
                },
              ],
              "title": "B",
            },
            Object {
              "suggestions": Array [
                Object {
                  "title": "Cherry",
                },
              ],
              "title": "C",
            },
          ]
    }
    noResultText="No result."
    onChange={[Function]}
    onSelect={[Function]}
    position="left"
    searching={false}
    searchingText="Searching for matches…"
    value="" />,
  "nodes": Array [
    <Typeahead
      id={42}
      items={
            Array [
                  Object {
                    "suggestions": Array [
                      Object {
                        "title": "Apple",
                      },
                      Object {
                        "title": "Apricot",
                      },
                    ],
                    "title": "A",
                  },
                  Object {
                    "suggestions": Array [
                      Object {
                        "title": "Banana",
                      },
                    ],
                    "title": "B",
                  },
                  Object {
                    "suggestions": Array [
                      Object {
                        "title": "Cherry",
                      },
                    ],
                    "title": "C",
                  },
                ]
      }
      noResultText="No result."
      onChange={[Function]}
      onSelect={[Function]}
      position="left"
      searching={false}
      searchingText="Searching for matches…"
      value="" />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <StringCompletionWidget
        fetchItems={[Function]}
        options={
                Object {
                        "itemsSrc": "/test.json",
                      }
        } />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "fetchItems": [Function],
          "options": Object {
            "itemsSrc": "/test.json",
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Typeahead
          id={42}
          items={
                    Array [
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Apple",
                                  },
                                  Object {
                                    "title": "Apricot",
                                  },
                                ],
                                "title": "A",
                              },
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Banana",
                                  },
                                ],
                                "title": "B",
                              },
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Cherry",
                                  },
                                ],
                                "title": "C",
                              },
                            ]
          }
          noResultText="No result."
          onChange={[Function]}
          onSelect={[Function]}
          position="left"
          searching={false}
          searchingText="Searching for matches…"
          value="" />,
        "_debugID": 2,
        "_renderedOutput": <Typeahead
          id={42}
          items={
                    Array [
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Apple",
                                  },
                                  Object {
                                    "title": "Apricot",
                                  },
                                ],
                                "title": "A",
                              },
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Banana",
                                  },
                                ],
                                "title": "B",
                              },
                              Object {
                                "suggestions": Array [
                                  Object {
                                    "title": "Cherry",
                                  },
                                ],
                                "title": "C",
                              },
                            ]
          }
          noResultText="No result."
          onChange={[Function]}
          onSelect={[Function]}
          position="left"
          searching={false}
          searchingText="Searching for matches…"
          value="" />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <StringCompletionWidget
    fetchItems={[Function]}
    options={
        Object {
            "itemsSrc": "/test.json",
          }
    } />,
}
`;

exports[`StringCompletionWidget should render without items 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Typeahead
    id={42}
    items={Array []}
    noResultText="No result."
    onChange={[Function]}
    onSelect={[Function]}
    position="left"
    searching={false}
    searchingText="Searching for matches…"
    value="" />,
  "nodes": Array [
    <Typeahead
      id={42}
      items={Array []}
      noResultText="No result."
      onChange={[Function]}
      onSelect={[Function]}
      position="left"
      searching={false}
      searchingText="Searching for matches…"
      value="" />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <StringCompletionWidget
        fetchItems={[Function]}
        options={Object {}} />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "fetchItems": [Function],
          "options": Object {},
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Typeahead
          id={42}
          items={Array []}
          noResultText="No result."
          onChange={[Function]}
          onSelect={[Function]}
          position="left"
          searching={false}
          searchingText="Searching for matches…"
          value="" />,
        "_debugID": 4,
        "_renderedOutput": <Typeahead
          id={42}
          items={Array []}
          noResultText="No result."
          onChange={[Function]}
          onSelect={[Function]}
          position="left"
          searching={false}
          searchingText="Searching for matches…"
          value="" />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <StringCompletionWidget
    fetchItems={[Function]}
    options={Object {}} />,
}
`;
